### DISCUSSION TITLE
Linear cartesian path planning

### DISCUSSION THREAD

**Discussion by joao-pm-santos96** [tesseract-robotics/trajopt#336] [Q&A] [answered]
Hi developers!
This package really seams to achieve wonderful results! Congratulations!

I have one question: is it possible to build a plan where the TCP/end-effector follows a linear trajectory?

Thank you!

**Comment by marip8** [comment, marked as answer]
Nominally yes. Trajopt can accept joint waypoints or Cartesian waypoints, and it represents movement between these waypoints in joint space (i.e. a joint-interpolated arc, rather than a straight linear line). However, if your waypoints are spaced closely enough together and are on your linear trajectory, the joint-interpolated arc between waypoints approximates a line.

If you're using `trajopt` through `tesseract`, you can denote the `MoveInstructions` of your motion planning input as `LINEAR` moves; then the `SimplePlanner` (which generates a seed trajectory for Trajopt) will do linear interpolation between those waypoints at some specifiable density to approximate a line. Then Trajopt will optimize the trajectory per the other constraints that were added to the optimization

`trajopt_ifopt` also has a [Cartesian line constraint](https://github.com/tesseract-robotics/trajopt/blob/master/trajopt_ifopt/include/trajopt_ifopt/constraints/cartesian_line_constraint.h) that you could check out. I'm not entirely sure how it works or if it has been thoroughly tested though

**Reply by joao-pm-santos96** [reply]
Thank you for the detailed explanation! That is beautiful!

**Reply by joao-pm-santos96** [reply]
Also, if you may help, where can I find the list the possible constraints and/or costs that we can add to the request?

**Reply by marip8** [reply]
The [headers in `trajopt_ifopt`](https://github.com/tesseract-robotics/trajopt/tree/master/trajopt_ifopt/include/trajopt_ifopt/constraints) provide a conceptual overview of what constraints are available. I think the only one that hasn't been ported from `trajopt` is the singularity avoidance constraint. 

The `tesseract` [default trajopt composite profile](https://github.com/tesseract-robotics/tesseract_planning/blob/master/tesseract_motion_planners/trajopt/include/tesseract_motion_planners/trajopt/profile/trajopt_default_composite_profile.h) and [plan profile](https://github.com/tesseract-robotics/tesseract_planning/blob/master/tesseract_motion_planners/trajopt/include/tesseract_motion_planners/trajopt/profile/trajopt_default_plan_profile.h#L54-L56) also set up some default constraints that you can enable/disable and set different cost values for
